File: CodeForces/N.py
import math


kx, ky = map(int, input().split())
n = int(input())
count = {x: {} for x in range(1, kx + 1)}
count_x = {x: 0 for x in range(1, kx + 1)}
x_y = {x: set() for x in range(1, kx + 1)}

for i in range(n):
    x, y = map(int, input().split())
    count_x[x] += 1
    if y in x_y[x]:
        count[x][y] += 1
    else:
        count[x][y] = 1
        x_y[x].add(y)

s = 0.
for x in range(1, kx + 1):
    ps = 0.
    for y in x_y[x]:
        temp = count[x][y] / count_x[x]
        ps -= temp * math.log(temp)
    s += count_x[x] / n * ps

print(s)

File: CF PABC.NET/N.pas
begin
  var (KX, KY) := ReadlnInteger2();
  var N := ReadlnInteger();
  
  var Count := Range(1, KX).ToDictionary(K -> K, V -> new Dictionary<Integer, Integer>);
  var CountX := Range(1, KX).ToDictionary(K -> K, V -> 0);
  var XY := Range(1, KX).ToDictionary(K -> K, V -> new HashSet<Integer>());
  
  loop N do
  begin
    var (X, Y) := ReadlnInteger2();
    CountX[X] += 1;
    if XY[X].Contains(Y) then
      Count[X][Y] += 1
    else begin
      Count[X].Add(Y, 1);
      XY[X].Add(Y);
    end;
  end;
  
  var S := 0.0;
  for var X := 1 to KX do
  begin
    var PS := 0.0;
    foreach var Y in XY[X] do
    begin
      var Temp := Count[X][Y] / CountX[X];
      PS -= Temp * Log(Temp);
    end;
    S += CountX[X] / N * PS;
  end;
  
  S.Println();
end.